### PR TITLE
feat: adding secondary label support

### PR DESCRIPTION
**Pull request by henri-vdm**
I think this a feature that could really benefit the flexibility of this package, it plays well with existing customLabels and you can mix match them. It’s a feature that I’ve found lacking when trying to use this library.

### CODE REVIEW

**Comment by henri-vdm** [comment]
I’ll add the documentation and storybook addition when if this PR goes through.

File: src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx [hunk 1]
```diff
@@ -70,6 +70,18 @@ describe("ReactFlagsSelect", () => {
         expect(screen.getByText("GB")).toBeInTheDocument();
       });
 
+      it("renders a special custom label of the valid selected country if present", () => {
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx:73]: Can we rephrase this to emphasize that we are testing for a custom label with value as object.
for example - `renders a custom label of the valid selected country if present as a CustomLabel object`.
Hence the previous test should be `... if present as a string`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx [hunk 2]
```diff
@@ -81,6 +93,18 @@ describe("ReactFlagsSelect", () => {
...
+      it("does not render a secondary label of the selected country if showSecondarySelectedLabel is false", () => {
+        render(
+          <ReactFlagsSelect
+            {...defaultProps}
+            selected="US"
+            showSecondarySelectedLabel={false}
+            customLabels={{ US: { primary: "US", secondary: "+1" } }}
+          />
+        );
+        expect(screen.queryByText("United States")).not.toBeInTheDocument();
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx:105]: This test does not check if the secondary label in the document.
It should also check if the primary label is in the document

File: src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx [hunk 3]
```diff
@@ -131,11 +155,29 @@ describe("ReactFlagsSelect", () => {
         expect(screen.getByText("US")).toBeInTheDocument();
       });
 
+      it("renders a custom label of each country option if present", () => {
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx:158]: This is kind of a repetition of the previous test above it (same description too).
I think you can use the previous test and modify one label to use the CustomLabel object

File: src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx [hunk 4]
```diff
@@ -131,11 +155,29 @@ describe("ReactFlagsSelect", () => {
...
+        expect(screen.getByText("+44")).toBeInTheDocument();
+        expect(screen.getByText("US")).toBeInTheDocument();
+      });
+
       it("does not render the label of each country option if showOptionLabel is false", () => {
         openOptions({ showOptionLabel: false });
         expect(screen.queryByText("Nigeria")).not.toBeInTheDocument();
       });
 
+      it("does not render the label of each country option if showSecondaryOptionLabel is false", () => {
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx:173]: `does not render the secondary label ...`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx [hunk 5]
```diff
@@ -131,11 +155,29 @@ describe("ReactFlagsSelect", () => {
...
+        openOptions({
+          showSecondaryOptionLabel: false,
+          customLabels: { US: { primary: "US", secondary: "+1" } },
+        });
+        expect(screen.queryByText("+1")).not.toBeInTheDocument();
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.test.tsx:178]: test that the primary label renders

File: src/components/ReactFlagsSelect/ReactFlagsSelect.tsx [hunk 6]
```diff
@@ -103,6 +107,9 @@ const ReactFlagsSelect: React.FC<Props> = ({
 
     const filteredCountriesOptions = countriesOptions.filter((key) => {
       const label = getLabel(key);
+      if (typeof label === "object") {
+        return label && label.primary.match(new RegExp(value, "i"));
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:111]: On a second thought, let's use optional chaining here `label?.primary?.match(`
Please help update the return statement below to use optional chaining too 🙏🏾

Comment (ekwonye-richard) [reply src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:111]: I think we should also search the secondary value -
`label?.primary?.match(new RegExp(value, "i")) || label?.secondary?.match(new RegExp(value, "i"))`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.tsx [hunk 7]
```diff
@@ -103,6 +107,9 @@ const ReactFlagsSelect: React.FC<Props> = ({
 
     const filteredCountriesOptions = countriesOptions.filter((key) => {
       const label = getLabel(key);
+      if (typeof label === "object") {
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:110]: I see we are doing `typeof label === "object"` quite a lot.
Let's expose a utility function in the `/utils` folder, maybe call it `isCustomLabelObject`.
This function can go a further step check if the object has at least the `primary` value.

File: src/components/ReactFlagsSelect/ReactFlagsSelect.tsx [hunk 8]
```diff
@@ -189,9 +198,17 @@ const ReactFlagsSelect: React.FC<Props> = ({
...
               {showSelectedLabel && (
                 <span className={styles.label}>
-                  {getLabel(validSelectedValue)}
+                  {typeof displayLabel === "object"
+                    ? displayLabel.primary
+                    : displayLabel}
                 </span>
               )}
+              {showSecondarySelectedLabel &&
+                typeof displayLabel === "object" && (
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:207]: The customLabel object should have the secondary value as optional, let's check if the secondary value is available before rendering this span.
Also the object check should use the utility function `isCustomLabelObject(displayLabel)`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.tsx [hunk 9]
```diff
@@ -189,9 +198,17 @@ const ReactFlagsSelect: React.FC<Props> = ({
...
+                  <span className={[styles.label, styles.secondary].join(" ")}>
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:208]: we are using `classnames` library to merge classes in this project.
`cx(styles.label, styles.secondaryLabel)`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.module.scss [hunk 10]
```diff
@@ -61,6 +62,10 @@ $gray-alt: #bdbbbb;
   padding-left: 10px;
   overflow: hidden;
   text-overflow: ellipsis;
+
+  &.secondary {
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.module.scss:66]: let's call this `secondaryLabel`

File: src/components/ReactFlagsSelect/ReactFlagsSelect.tsx [hunk 11]
```diff
@@ -245,9 +263,19 @@ const ReactFlagsSelect: React.FC<Props> = ({
...
-                      {getLabel(countryCode)}
+                      {typeof countryLabel === "object"
+                        ? countryLabel.primary
+                        : countryLabel}
                     </span>
                   )}
+                  {showSecondaryOptionLabel &&
+                    typeof countryLabel === "object" && (
+                      <span
+                        className={[styles.label, styles.secondary].join(" ")}
```
Comment (ekwonye-richard) [inline src/components/ReactFlagsSelect/ReactFlagsSelect.tsx:274]: let's use `cx`

File: src/types/ReactFlagsSelectTypes.ts [hunk 12]
```diff
@@ -1,4 +1,6 @@
 export type Countries = Record<string, string>;
 export type CountryCodes = Array<string>;
-export type CustomLabels = Record<string, string>;
+export type CustomLabels =
+  | Record<string, string>
+  | Record<string, { primary: string; secondary: string }>;
```
Comment (ekwonye-richard) [inline src/types/ReactFlagsSelectTypes.ts:5]: let's export a type called CustomLabel which is essentially
`{ 
    primary: string;
   secondary?: string
}`
I'll suggest secondary been optional

Comment (ekwonye-richard) [reply src/types/ReactFlagsSelectTypes.ts:5]: Hence `customLabels = Record<string, CustomLabel | string>`

**Comment by ekwonye-richard** [comment]
Hi @henri-vdm thanks a lot for introducing this very thoughtful feature into this library. I really appreciate your contribution and I'm very sure the community will benefit from this feature.

It will be nice to have the feature in one PR so you should update storybook and the README with your changes.
It will also be great if you can attach a screenshots to this PR.

Thanks @henri-vdm 🎉

**Comment by henri-vdm** [comment]
> Hi @henri-vdm thanks a lot for introducing this very thoughtful feature into this library. I really appreciate your contribution and I'm very sure the community will benefit from this feature.
> 
> It will be nice to have the feature in one PR so you should update storybook and the README with your changes.
> It will also be great if you can attach a screenshots to this PR.
> 
> Thanks @henri-vdm 🎉

All good made the changes and add the storybook and readme changes.

Some pics:
demo1
demo2

**Review by ekwonye-richard — APPROVED** [review]

**Comment by ekwonye-richard** [comment]
Thanks you @henri-vdm for your contribution 🎊
